parasolr/django/signals.py

```diff
@@ -61,89 +61,86 @@ class MyModel(Model, ModelIndexable):
-                                 model)
-                    models.signals.pre_save.disconnect(options['save'],
-                                                       sender=model)
-                if 'delete' in options:
-                    logger.debug('Disconnecting delete signal handler for %s',
-                                 model)
-                    models.signals.pre_delete.disconnect(
-                        options['delete'], sender=model)
+    @staticmethod
+    def connect():
+        '''bind indexing signal handlers to save and delete signals for
+        :class:`~ppa.archive.solr.Indexable` subclassess and any
+        indexing dependencies'''
+
+        # bind to save and delete signals for ModelIndexable subclasses
+        for model in ModelIndexable.__subclasses__():
+            logger.debug('Registering signal handlers for %s', model)
+            models.signals.post_save.connect(
+                IndexableSignalHandler.handle_save, sender=model)
+            models.signals.post_delete.connect(
+                IndexableSignalHandler.handle_delete, sender=model)
+
+        ModelIndexable.identify_index_dependencies()
+        for m2m_rel in ModelIndexable.m2m:
+            logger.debug('Registering m2m signal handler for %s', m2m_rel)
+            models.signals.m2m_changed.connect(
+                IndexableSignalHandler.handle_relation_change,
+                sender=m2m_rel)
+
+        for model, options in ModelIndexable.related.items():
+            for signal_name, handler in options.items():
+                model_signal = getattr(models.signals, signal_name)
+                logger.debug('Registering %s signal handler for %s',
+                             signal_name, model)
+                model_signal.connect(handler, sender=model)
+
+    @staticmethod
+    def disconnect():
+        '''disconnect indexing signal handlers'''
+        for model in ModelIndexable.__subclasses__():
+            logger.debug('Disconnecting signal handlers for %s', model)
+            models.signals.post_save.disconnect(
+                IndexableSignalHandler.handle_save, sender=model)
+            models.signals.post_delete.disconnect(
+                IndexableSignalHandler.handle_delete, sender=model)
+
+        for m2m_rel in ModelIndexable.m2m:
+            logger.debug('Disconnecting m2m signal handler for %s',
+                         m2m_rel)
+            models.signals.m2m_changed.disconnect(
+                IndexableSignalHandler.handle_relation_change,
+                sender=m2m_rel)
+
+        for model, options in ModelIndexable.related.items():
+            for signal_name, handler in options.items():
+                model_signal = getattr(models.signals, signal_name)
+                logger.debug('Disconnecting %s signal handler for %s',
+                             signal_name, model)
+                model_signal.disconnect(handler, sender=model)
 
+
+if django:
```

Comment:
can the decorator be used here too?

Reply:
No, because you can't decorate a function call.

We could change it so it doesn't auto-connect, and you have to call connect in your apps ready method, though.

Reply:
ah, only function/class declarations? makes sense. i'm fine with this for now